File: src/editor/database/card/MeldCard.java
package editor.database.card;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import javax.swing.text.BadLocationException;
import javax.swing.text.Style;
import javax.swing.text.StyledDocument;

import editor.database.characteristics.ManaCost;
import editor.util.Lazy;

/**
 * This class represents a "meld" card, which is a double-faced card where the back face is
 * only half of a card (with the other half belonging to another card).  It knows about its
 * "sibling" for deriving characteristics of the back face but doesn't display any of its
 * "sibling's" information.
 *
 * @author Alec Roelke
 */
public class MeldCard extends MultiCard
{
    /**
     * Converted mana costs of this MeldCard's faces.
     */
    private Lazy<List<Double>> cmc;
    /**
     * Front face of this MeldCard.
     */
    private final Card front;
    /**
     * Tuple of this MeldCard's faces' mana costs.
     */
    private Lazy<List<ManaCost>> manaCost;
    /**
     * This MeldCard's "sibling," with which it melds to form the back face.
     */
    private final Card other;

    /**
     * Create a new MeldCard, with the given cards as front, other front, and
     * back faces, respectively.
     *
     * @param f front face of the new MeldCard
     * @param o other front face of the new MeldCard
     * @param b back face of the two front faces.
     */
    public MeldCard(Card f, Card o, Card b)
    {
        super(CardLayout.MELD, f, b);
        front = f;
        other = o;

        if (front.layout() != CardLayout.MELD || other.layout() != CardLayout.MELD || b.layout() != CardLayout.MELD)
            throw new IllegalArgumentException("can't join non-meld cards into meld cards");

        manaCost = new Lazy<>(() -> Collections.unmodifiableList(Arrays.asList(front.manaCost().get(0), new ManaCost())));
        cmc = new Lazy<>(() -> Collections.unmodifiableList(Arrays.asList(front.cmc().get(0), front.cmc().get(0) + other.cmc().get(0))));
    }

    /**
     * {@inheritDoc}
     * While only the front face has a mana cost, the back face's converted mana cost
     * is the sum of those of its two front faces.
     */
    @Override
    public List<Double> cmc()
    {
        return cmc.get();
    }

    /**
     * {@inheritDoc}
     * Includes some information to the front face indicating what the other front face is.
     */
    @Override
    public void formatDocument(StyledDocument document, int f)
    {
        Style reminderStyle = document.getStyle("reminder");
        super.formatDocument(document, f);
        if (f == 0)
        {
            try
            {
                document.insertString(document.getLength(), "\nMelds with " + other.unifiedName(), reminderStyle);
            }
            catch (BadLocationException e)
            {
                e.printStackTrace();
            }
        }
    }

    /**
     * {@inheritDoc}
     * The back face of a MeldCard has no mana cost (but does have a converted mana cost,
     * see {@link MeldCard#cmc()}).
     */
    @Override
    public List<ManaCost> manaCost()
    {
        return manaCost.get();
    }

    @Override
    public double minCmc()
    {
        return front.minCmc();
    }
}

File: src/main/scala/editor/database/card/MeldCard.scala
package editor.database.card

import editor.database.attributes.ManaCost

import javax.swing.text.BadLocationException
import javax.swing.text.StyledDocument

import CardLayout.MELD

/**
 * One half of a meld card, which is a card that can combine with another during a game to form
 * another card represented by the back halves of the two combined cards. It knows about the other
 * card that forms the other half of the back, but doesn't display any information about it. The
 * mana value of a meld card is that of the front face only.
 * 
 * @constructor create a new meld card
 * @param front front face of the card
 * @param other front face of the other card that forms the back
 * @param back fully-formed back face of the two halves
 * @author Alec Roelke
 */
@throws[IllegalArgumentException]("if any of the faces aren't part of a meld card")
class MeldCard(front: Card, other: Card, back: Card) extends MultiCard(MELD, IndexedSeq(front, back)) {
  if (front.layout != MELD || other.layout != MELD || back.layout != MELD)
    throw IllegalArgumentException("can't join non-meld cards into meld cards")
  
  override def manaCost = front.manaCost

  override def manaValue = front.manaValue
  override def minManaValue = faces.map(_.manaValue).min
  override def maxManaValue = faces.map(_.manaValue).max
  override def avgManaValue = faces.map(_.manaValue).sum/faces.size
  override lazy val imageNames = front.imageNames ++ back.imageNames

  override def formatDocument(document: StyledDocument, printed: Boolean, f: Int) = {
    val reminderStyle = document.getStyle("reminder");
    super.formatDocument(document, printed, f);
    if (f == 0) {
      try {
          document.insertString(document.getLength(), s"\nMelds with ${other.name} (${other.expansion})", reminderStyle);
      } catch case e: BadLocationException => e.printStackTrace
    }
  }
}